### PR TITLE
Add beam utility for finding reachability of a graph

### PR DESCRIPTION
**Pull request by wchamber**
Use JUnit 4 in workflow tests as the Apache Beam TestPipeline makes use of it.
To be used to make hierarchical relationship indexes job in another PR.

### CODE REVIEW

File: workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+      PCollection<KV<T, T>> nMinus1AllPaths = allPaths.get(n - 1);
+
+      PCollection<KV<T, T>> newAllPaths =
+          PCollectionList.of(nMinus1AllPaths)
+              .and(nExactPaths)
+              .and(concatenate(nExactPaths, nMinus1AllPaths, "allPaths temp N" + n))
+              .apply("flatten N" + n, Flatten.pCollections());
+      int newLongestPathLength = 2 * n - 1;
+      if (newLongestPathLength >= maxPathLength) {
```
Comment (calbach) [inline workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java:81]: nit: maybe rename `maxPathLength` to make it clearer that this is to bound computation, and is not an exact cutoff to the answer e.g. `maxPathLengthHint`?

File: workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    Map<Integer, PCollection<KV<T, T>>> allPaths = new HashMap<>();
+    // exactPaths[n] is the set of vertices where there exists a path from V0 -> V1 in exactly N
+    // edges.
+    Map<Integer, PCollection<KV<T, T>>> exactPaths = new HashMap<>();
+
+    // The input edges are both allPaths[1] and exactPaths[1].
+    allPaths.put(1, edges);
+    exactPaths.put(1, edges);
+
+    // Iteratively build up allPaths in O(log_2(maxPathLength)) joins. We want to avoid doing many
```
Comment (calbach) [inline workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java:51]: Nice algorithm. Is there a name for this or was this something you just fit to purpose for this? May be worth linking a wikipedia article or something for visual learners, if there is any such documentation

File: workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+      PCollection<KV<T, T>> newAllPaths =
+          PCollectionList.of(nMinus1AllPaths)
+              .and(nExactPaths)
+              .and(concatenate(nExactPaths, nMinus1AllPaths, "allPaths temp N" + n))
+              .apply("flatten N" + n, Flatten.pCollections());
+      int newLongestPathLength = 2 * n - 1;
+      if (newLongestPathLength >= maxPathLength) {
+        // Stop when we have computes at least maxPathLength length paths.
+        return newAllPaths;
```
Comment (calbach) [inline workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java:83]: opt nit: technically we could also exit when the output set stops changing, but this operation might be the same cost or similar to just running out the rest of the cycles, especially since it's a logarithmic number of steps.

File: workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.TupleTag;
+
+/** Utilities for working with graphs in Apache Beam. */
+public final class GraphUtils {
+  private GraphUtils() {}
+
+  /**
+   * Computes the transitive closure of a directed acyclic graph defined by {@code edges}.
```
Comment (calbach) [inline workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java:25]: Out of curiosity, what happens if this is not acyclic? Is that actually a prerequisite of this function?

File: workflow/src/test/java/bio/terra/tanagra/workflow/GraphUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  /** Convert a multimap to an equivalent list of KVs. */
+  private static <T> List<KV<T, T>> convertToKvs(Multimap<T, T> multimap) {
+    return multimap.entries().stream()
+        .map(entry -> KV.of(entry.getKey(), entry.getValue()))
+        .collect(Collectors.toList());
+  }
+
+  @Test
```
Comment (calbach) [inline workflow/src/test/java/bio/terra/tanagra/workflow/GraphUtilsTest.java:40]: opt: may be worth including some smaller edge cases for the logarithmic approach, e.g. single-edge graph, with a maxPathLen of 1

**Review by calbach — APPROVED** [review]
Nice!

File: workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java (hunk 4 above)
Comment (wchamber) [reply workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java:25]: Nope, though it will generate a lot of duplicates. Might want to add calls to `Distinct` to deduplicate if that happens often.

File: workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java (hunk 3 above)
Comment (wchamber) [reply workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java:83]: Beam doesn't support that "conditional iteration" yet, so unfortunately we can't exit when the output set doesn't change. Added a comment and link

File: workflow/src/test/java/bio/terra/tanagra/workflow/GraphUtilsTest.java (hunk 5 above)
Comment (wchamber) [reply workflow/src/test/java/bio/terra/tanagra/workflow/GraphUtilsTest.java:40]: Good idea, added.

File: workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java (hunk 2 above)
Comment (wchamber) [reply workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java]: Found internally in a google doc initially, but added a link to a paper reference. No great graphics unfortunately.

File: workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java (hunk 1 above)
Comment (wchamber) [reply workflow/src/main/java/bio/terra/tanagra/workflow/GraphUtils.java]: thanks, done.
